h/static/scripts/directive/annotation.js

```diff
@@ -72,50 +89,93 @@ function extractDocumentMetadata(model) {
+ * saved to the server yet.
+ */
+function isNew(domainModel) {
+  return !domainModel.id;
+}
 
-  // If we have unsaved changes to this annotation, apply them
-  // to the view model.
+/** Restore unsaved changes to this annotation from the drafts service.
+ *
+ * If there are no draft changes to this annotation, does nothing.
+ *
+ */
+function restoreFromDrafts(drafts, domainModel, vm) {
+  var draft = drafts.get(domainModel);
   if (draft) {
-    angular.extend(vm.annotation, angular.copy(draft));
+    vm.isPrivate = draft.isPrivate;
+    vm.form.tags = draft.tags;
+    vm.form.text = draft.text;
   }
+}
 
-  vm.annotationURI = new URL(
-    '/a/' + vm.annotation.id, vm.baseURI).href;
+/**
+  * Save the given annotation to the drafts service.
+  *
+  * Any existing drafts for this annotation will be overwritten.
+  *
+  * @param {object} drafts - The drafts service
+  * @param {object} domainModel - The full domainModel object of the
+  *   annotation to be saved. This full domainModel model is not retrieved
+  *   again from drafts, it's only used to identify the annotation's draft in
+  *   order to retrieve the fields below.
+  * @param {object} vm - The view model object containing the user's unsaved
+  *   changes to the annotation.
+  *
+  */
+function saveToDrafts(drafts, domainModel, vm) {
+  drafts.update(
+    domainModel,
+    {
+      isPrivate: vm.isPrivate,
+      tags: vm.form.tags,
+      text: vm.form.text,
+    });
+}
 
-  vm.document = extractDocumentMetadata(model);
+/** Update domainModel from vm.
+ *
+ * Copy any properties from vm that might have been modified by the user into
+ * domainModel, overwriting any existing properties in domainModel.
+ *
+ * Additionally, the `tags` property of vm - an array of objects each with a
+ * `text` string - will become a simple array of strings in domainModel.
+ *
+ * @param {object} domainModel The object to copy properties to
+ * @param {object} vm The object to copy properties from
+ * @returns undefined
+ *
+ */
+function updateDomainModel(domainModel, vm, permissions, groups) {
+  domainModel.text = vm.form.text;
+  domainModel.tags = domainModelTagsFromViewModelTags(vm.form.tags);
+  if (vm.isPrivate) {
+    domainModel.permissions = permissions.private();
+  } else {
+    domainModel.permissions = permissions.shared(domainModel.group);
+  }
+  domainModel.group = groups.focused().id;
+}
 
-  // Form the tags for ngTagsInput.
-  vm.annotation.tags = (vm.annotation.tags || []).map(function(tag) {
-    return {text: tag};
-  });
+/** Update the view model from the domain model changes. */
+function updateViewModel(domainModel, vm, permissions) {
+  // Extend the view model with a copy of the domain model.
+  // Note that copy is used so that deep properties aren't shared.
+  vm.form = {
+    text: domainModel.text,
+    tags: viewModelTagsFromDomainModelTags(domainModel.tags),
+  };
+  vm.annotationURI = new URL('/a/' + domainModel.id, vm.baseURI).href;
```

Comment:
Why are we constructing the annotation link ourselves here, doesn't the domainModel include a property which is the link? (ie. `domainModel.link`?)